units/log_output_tests.cc

```diff
@@ -0,0 +1,141 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cstddef>
+#include <cstdio>
+#include <fstream>
+#include <string>
+#include <thread>
+
+#include "debug_logging.h"
+#include "gtest/gtest.h"
+
+#define LOGS_PER_THREAD 1337
+#define THREAD_COUNT 100
+
+namespace performancelayers {
+namespace {
+// An abstraction over the output. It writes the incoming string to the output.
+class LogOutput {
+ public:
+  virtual ~LogOutput() = default;
+
+  virtual void LogLine(const std::string &line) = 0;
+};
+
+// Implements LogOutput for a file. If the given filename is `nullptr`, it
+// writes to the standard output. After each write, the logs are persisted to
+// the file by calling `fflush`. Since only one line is written to the output by
+// each `LogLine()` call, it doens't need to aquire a mutex.
+class FileOutput : public LogOutput {
+ public:
+  FileOutput(const char *filename) {
+    if (filename) {
+      // Since multiple layers open the same file and write to it at the same
+      // time, it's opened in the append mode.
+      out_ = fopen(filename, "a");
+      if (!out_) {
+        SPL_LOG(ERROR) << "Failed to open " << filename
+                       << ". Using stderr as the alternative output.";
+        out_ = stderr;
+      }
+    } else {
+      out_ = stderr;
+    }
+  }
+
+  ~FileOutput() {
+    if (out_ && out_ != stderr) {
+      fclose(out_);
+      out_ = nullptr;
```

Comment:
nit: we can't run any other code after the destructor so we don't have to zero-out values. Feel free to keep it if you found it useful when debugging etc